### PR TITLE
fix: validate Pyth confidence interval before caching prices

### LINKED ISSUE #67: [HIGH-CRITICAL] bug: Pyth confidence interval is ignored before trusting oracle prices
## Summary

The keeper currently accepts fresh positive Pyth prices without validating the Pyth confidence interval (`conf`).

`fetchPythPrices()` reads and uses:

```txt
price.price
price.expo
price.publish_time
```

but does not read or validate:

```txt
price.conf
```

As a result, a Pyth price with an extremely wide confidence interval can still be cached, returned as `source="pyth"`, and treated as the highest-trust price source.

This matters because `source="pyth"` also bypasses the first-push secondary-source cross-check.

## Non-duplicate note

This is not the same as the previous Pyth/Hermes transport issue.

That issue was about:

```txt
HERMES_URL HTTPS validation
NODE_TLS_REJECT_UNAUTHORIZED=0
Pyth Hermes binary VAA presence
transport / attestation presence
```

This report is about a different issue:

```txt
Pyth price quality
confidence interval validation
accepting a degraded / low-confidence Pyth price
bypassing first-push secondary confirmation because source="pyth"
```

So this is not a duplicate of the Hermes/TLS/VAA issue. That issue ensures the response is transported and attestation-backed. This issue checks whether the accepted Pyth price is precise enough to trust as an oracle price.

## Impact

This is an oracle correctness issue.

Pyth prices include a confidence interval that represents uncertainty around the reported price. A very large confidence interval means the feed is degraded or the price is highly uncertain.

Current keeper behavior can accept such a price if:

```txt
price > 0
publish_time is fresh
binary.data is present
```

Potential impact:

* degraded Pyth prices can be cached as valid prices
* a low-confidence Pyth price can become the keeper's primary price source
* `source="pyth"` bypasses the first-push cross-check
* after restart, the first accepted Pyth price can become the local baseline
* circuit-breaker behavior may compare later prices against an unsafe baseline
* operators may assume the price is high-trust because it came from Pyth, even though the confidence interval is extremely wide

Severity: High-Critical / Critical candidate.

The issue is high-critical because Pyth is treated as the most trusted source in the keeper. If a degraded Pyth price with a very wide confidence interval is accepted, it can become the oracle baseline without secondary confirmation.

## Root Cause

The current Pyth parsing logic validates freshness and positive price, but does not validate confidence.

Current behavior:

```ts
const rawPrice = parseInt(entry.price.price, 10);
const expo = entry.price.expo;
const price = rawPrice * Math.pow(10, expo);

const publishMs = entry.price.publish_time * 1000;
const ageMs = Date.now() - publishMs;

if (price > 0 && ageMs >= 0 && ageMs < 60_000) {
  pythCache.set(sym, { price, ts: publishMs });
}
```

There is no check for:

```ts
entry.price.conf
```

Then `getPrice()` treats cached Pyth prices as primary:

```ts
const pyth = getPythPrice(symbol);
if (pyth) return { price: pyth.price, source: "pyth", freshAt: pyth.freshAt };
```

And first-push cross-check is skipped for Pyth:

```ts
if (s.lastPrice === 0 && source !== "pyth") {
  // secondary-source confirmation only runs for non-Pyth prices
}
```

This means a fresh but extremely low-confidence Pyth price can be accepted and trusted without secondary confirmation.

## Why this is unsafe

Pyth confidence interval is part of the price quality signal.

For example:

```txt
price = 150
conf  = 150
conf / price = 100%
```

A confidence interval equal to 100% of the price means the feed is extremely uncertain. The keeper should not treat this the same way as a precise Pyth price.

A safer keeper should reject or downgrade Pyth prices when:

```txt
conf / price is too high
conf is missing
conf is invalid
conf is non-finite
```

At minimum, a configurable threshold should exist, for example:

```txt
PYTH_MAX_CONFIDENCE_PCT=10
```

## Failure Scenario

1. Keeper starts or restarts.
2. Pyth Hermes returns a fresh positive price.
3. The response includes non-empty `binary.data`.
4. The Pyth price has an extremely wide confidence interval.
5. Current keeper code ignores `conf`.
6. The price is cached as valid.
7. `getPrice()` returns it as `source="pyth"`.
8. First-push secondary-source cross-check is skipped because the source is Pyth.
9. The low-confidence price can become the keeper's local baseline.

## PoC

The following PoC mirrors the current keeper behavior:

1. Pyth response contains a fresh positive price.
2. `binary.data` is present.
3. `conf` is extremely wide: 100% of price.
4. Current logic ignores `conf`.
5. The price is cached.
6. `getPrice()` returns `source="pyth"`.
7. First-push cross-check does not run.
8. The assertion fails because the low-confidence price should not be cached.

```js
const assert = require('assert');

// Mirrors current keeper behavior:
// - fetchPythPrices() only reads price.price, price.expo, and price.publish_time.
// - It does not read or validate price.conf.
// - A fresh positive Pyth price is cached even if the confidence interval is extremely wide.
// - getPrice() then treats it as source="pyth".
// - First-push cross-check is skipped for source="pyth".

const PYTH_FEED_IDS = {
  SOL: 'ef0d8b6fda2ceba41da15d4095d1da392a0d2f8ed0c6c7bc0f4cfac8c280b56d',
};

const pythCache = new Map();

const nowSec = Math.floor(Date.now() / 1000);

const hermesResponseWithHugeConfidence = {
  binary: {
    encoding: 'base64',
    data: ['fake-non-empty-vaa-presence-check-only'],
  },
  parsed: [
    {
      id: PYTH_FEED_IDS.SOL,
      price: {
        // price = 150
        price: '15000000',
        expo: -5,

        // conf = 150, i.e. confidence interval is 100% of price.
        // Current keeper behavior ignores this field completely.
        conf: '15000000',

        publish_time: nowSec,
      },
    },
  ],
};

// Mirrors current fetchPythPrices() behavior.
function fetchPythPrices_likeCurrentCode(json) {
  if (!json.binary?.data || json.binary.data.length === 0) {
    return;
  }

  const idToSymbol = new Map(
    Object.entries(PYTH_FEED_IDS).map(([symbol, id]) => [id, symbol])
  );

  for (const entry of json.parsed) {
    const sym = idToSymbol.get(entry.id);
    if (!sym) continue;

    const rawPrice = parseInt(entry.price.price, 10);
    const expo = entry.price.expo;
    const price = rawPrice * Math.pow(10, expo);

    const publishMs = entry.price.publish_time * 1000;
    const ageMs = Date.now() - publishMs;

    // Current code only checks price > 0 and freshness.
    // It does not check entry.price.conf at all.
    if (price > 0 && ageMs >= 0 && ageMs < 60_000) {
      pythCache.set(sym, { price, ts: publishMs });
    }
  }
}

function getPythPrice_likeCurrentCode(symbol) {
  const cached = pythCache.get(symbol);
  if (!cached) return null;
  if (Date.now() - cached.ts > 30_000) return null;
  return { price: cached.price, freshAt: cached.ts };
}

function getPrice_likeCurrentCode(symbol) {
  const pyth = getPythPrice_likeCurrentCode(symbol);
  if (pyth) return { price: pyth.price, source: 'pyth', freshAt: pyth.freshAt };
  return null;
}

function firstPushCrossCheckRuns_likeCurrentCode(lastPrice, source) {
  return lastPrice === 0 && source !== 'pyth';
}

// Expected safer behavior:
// reject Pyth prices when conf/price is too wide.
function shouldAcceptPythPrice_expectedSafe(entry, maxConfPct = 10) {
  const rawPrice = parseInt(entry.price.price, 10);
  const rawConf = parseInt(entry.price.conf, 10);
  const expo = entry.price.expo;

  const price = rawPrice * Math.pow(10, expo);
  const conf = rawConf * Math.pow(10, expo);
  const confPct = Math.abs(conf / price) * 100;

  return Number.isFinite(price) &&
    Number.isFinite(confPct) &&
    price > 0 &&
    confPct <= maxConfPct;
}

fetchPythPrices_likeCurrentCode(hermesResponseWithHugeConfidence);

const result = getPrice_likeCurrentCode('SOL');
const crossCheckRuns = firstPushCrossCheckRuns_likeCurrentCode(0, result?.source);
const expectedSafeAccept = shouldAcceptPythPrice_expectedSafe(
  hermesResponseWithHugeConfidence.parsed[0],
  10
);

console.log({
  pythParsedPrice: hermesResponseWithHugeConfidence.parsed[0].price,
  cachedPythPrice: pythCache.get('SOL') ?? null,
  result,
  firstPushCrossCheckRuns: crossCheckRuns,
  expectedSafeAccept,
  confidencePct: 100,
});

assert.strictEqual(
  pythCache.has('SOL'),
  false,
  'Pyth price with extremely wide confidence interval should not be cached'
);

assert.strictEqual(
  result,
  null,
  'getPrice should not return a Pyth price when confidence interval is too wide'
);

assert.strictEqual(
  crossCheckRuns,
  true,
  'unsafe low-confidence Pyth first push should not bypass secondary cross-check'
);
```

## Actual Result

```txt
{
  pythParsedPrice: {
    price: '15000000',
    expo: -5,
    conf: '15000000',
    publish_time: 1782438719
  },
  cachedPythPrice: {
    price: 150,
    ts: 1782438719000
  },
  result: {
    price: 150,
    source: 'pyth',
    freshAt: 1782438719000
  },
  firstPushCrossCheckRuns: false,
  expectedSafeAccept: false,
  confidencePct: 100
}

AssertionError [ERR_ASSERTION]: Pyth price with extremely wide confidence interval should not be cached

true !== false
```

## Expected Result

The keeper should reject or downgrade Pyth prices when the confidence interval is too wide.

Expected behavior for the PoC response:

```txt
confidencePct = 100
max allowed confidence = 10
result should not be cached
getPrice should not return source="pyth"
first-push should not bypass secondary confirmation
```

## Suggested Fix

Add confidence validation before caching Pyth prices.

Example direction:

```ts
const PYTH_MAX_CONFIDENCE_PCT = parsePositiveNumberEnv(
  "PYTH_MAX_CONFIDENCE_PCT",
  10,
  100,
);
```

Then validate `conf`:

```ts
const rawPrice = parseInt(entry.price.price, 10);
const rawConf = parseInt(entry.price.conf, 10);
const expo = entry.price.expo;

const price = rawPrice * Math.pow(10, expo);
const conf = rawConf * Math.pow(10, expo);
const confPct = Math.abs(conf / price) * 100;

if (
  price > 0 &&
  isFinite(price) &&
  isFinite(confPct) &&
  confPct <= PYTH_MAX_CONFIDENCE_PCT &&
  ageMs >= 0 &&
  ageMs < 60_000
) {
  pythCache.set(sym, { price, ts: publishMs });
} else if (price > 0 && confPct > PYTH_MAX_CONFIDENCE_PCT) {
  log(`⚠️ ${sym}: Pyth confidence too wide (${confPct.toFixed(2)}%) — rejecting price`);
}
```

If `conf` is missing, invalid, or too wide, the keeper should not treat the price as high-trust Pyth.

## Security Note

Pyth is the keeper's highest-trust source. That trust should include not only response freshness and VAA presence, but also price quality.

A fresh Pyth price with a very wide confidence interval should not be accepted as a strong oracle baseline without validation.

### PR DESCRIPTION
**Pull request by Bayyan16**
## Summary

Closes #67.

This PR fixes the Pyth oracle confidence validation issue by validating the Pyth confidence interval before any Pyth price is written into `pythCache`.

Previously, the keeper accepted a Pyth price when the price was positive and the `publish_time` was fresh, but it did not validate the `conf` field returned by Pyth. Because cached Pyth prices can later be used as trusted oracle data with `source: "pyth"`, a degraded Pyth update with a very wide confidence interval could still enter the keeper’s trusted price path.

This PR makes the Pyth price acceptance rule stricter and safer: a Pyth price must now have a valid confidence value, and the confidence interval must stay within the configured maximum threshold before the price can be cached.

---

## Background

Pyth price feeds provide more than just a raw price. A Pyth price update also includes a confidence interval, represented by `conf`, which indicates the uncertainty around the reported price.

Before this change, the keeper was already reading and using:

- `price.price`
- `price.expo`
- `price.publish_time`

However, it was not reading or validating:

- `price.conf`

That meant a Pyth price could be accepted if it was fresh and positive, even when the confidence interval was extremely wide. In practice, this weakens the safety assumptions around Pyth as a trusted source, especially during volatile market conditions, oracle degradation, thin liquidity, or temporary feed quality issues.

This PR closes that gap by enforcing confidence interval validation before caching the Pyth price.

---

## Root Cause

The issue was located in the Pyth price fetch and cache flow.

The previous flow accepted a Pyth price when:

1. the response contained usable Pyth data,
2. the raw price was greater than zero,
3. the `publish_time` was recent enough,
4. and then the computed price was stored into `pythCache`.

The missing part was that the keeper did not validate the uncertainty of that price. Specifically, it did not reject prices where `conf / price` was too large.

Because `pythCache` is later used by the keeper as a trusted Pyth source, the safest place to fix the issue is before `pythCache.set(...)`. This ensures low-confidence Pyth prices never enter the trusted cached price path.

---

### 1. Added Pyth confidence threshold configuration

A new environment variable was added:

- `PYTH_MAX_CONFIDENCE_PCT`

The default value is:

- `10`

This means a Pyth price is rejected when its confidence percentage is greater than the configured maximum threshold.

The confidence percentage is calculated as:

- `conf / price * 100`

This keeps the threshold explicit, configurable, and easy for maintainers to tune without changing code.

---

### 2. Added `src/pyth-confidence.ts`

A new helper module was added to keep Pyth price quality validation isolated and testable.

This helper is responsible for:

- parsing the raw Pyth price,
- parsing the raw Pyth confidence value,
- validating the exponent,
- applying the shared Pyth exponent to price and confidence,
- calculating the confidence percentage,
- rejecting invalid values,
- rejecting missing confidence,
- rejecting negative confidence,
- rejecting non-positive price values,
- rejecting non-finite computed values.

This helper was intentionally kept outside `src/index.ts` so it can be unit-tested without importing the full keeper runtime or triggering side effects from the main keeper process.

---

### 3. Updated Pyth response typing

The Pyth response type now includes the optional `conf` field.

This allows the keeper to explicitly inspect the confidence interval returned by Pyth instead of only reading `price`, `expo`, and `publish_time`.

---

### 4. Updated Pyth price cache flow

The Pyth fetch flow now validates confidence before caching the price.

The updated acceptance flow is:

1. Read the Pyth price update.
2. Parse `price`, `conf`, and `expo`.
3. Reject the update if price/conf data is missing or invalid.
4. Compute the final price.
5. Compute the confidence percentage.
6. Validate `publish_time`.
7. Reject the update if the confidence percentage is greater than `PYTH_MAX_CONFIDENCE_PCT`.
8. Reject the update if the Pyth update is stale.
9. Only then write the price into `pythCache`.

This is the most important part of the fix: `pythCache.set(...)` is only reached after the confidence interval has passed validation.

---

### 5. Preserved existing stale-price handling

This PR does not remove or weaken the existing freshness validation.

The existing stale-price behavior remains in place:

- invalid `publish_time` is rejected,
- future or stale updates are rejected,
- only fresh Pyth updates can be cached.

The new confidence validation is added on top of the existing freshness and positive-price checks.

---

### 6. Added tests for issue #67

New unit tests were added to cover the confidence validation behavior.

The tests cover:

- rejecting a fresh positive Pyth price when `conf / price` is too wide,
- accepting a Pyth price when confidence is within the configured threshold,
- accepting confidence exactly at the threshold,
- rejecting missing confidence,
- rejecting invalid confidence,
- rejecting negative confidence,
- rejecting zero price,
- rejecting invalid price values.

This makes the fix testable and prevents future regressions.

---

## Security Impact

This PR improves oracle safety by preventing degraded Pyth prices from becoming trusted cached data.

Before this fix, a fresh positive Pyth price could be cached even if the confidence interval was extremely wide. After this fix, the keeper requires the Pyth price to satisfy both freshness and confidence-quality checks before it can enter the cache.

This reduces the risk of accepting unreliable oracle data during:

- high volatility,
- temporary Pyth feed degradation,
- abnormal market conditions,
- wide uncertainty intervals,
- low-quality price updates.

The fix is intentionally minimal and targeted. It does not redesign the keeper, change source priority, remove fallback logic, or alter the broader circuit breaker behavior. It only hardens the Pyth acceptance path at the point where unsafe data would otherwise enter the cache.

---

### `.env.example`

Adds:

- `PYTH_MAX_CONFIDENCE_PCT=10`

This documents the new confidence threshold configuration.

### `src/index.ts`

Updates the Pyth fetch/cache flow to:

- import the Pyth confidence helper,
- read the Pyth `conf` field,
- validate Pyth price quality,
- reject invalid or missing confidence data,
- reject wide-confidence Pyth prices,
- cache only Pyth prices that pass all guards.

### `src/pyth-confidence.ts`

Adds a pure helper module for:

- parsing Pyth price payloads,
- calculating confidence percentage,
- validating price/conf correctness,
- exposing reusable validation logic for tests.

### `src/security-fixes.test.ts`

Adds regression tests for issue #67 covering valid, invalid, missing, and wide-confidence Pyth price scenarios.

---

## Tests

Validated locally with:

- `npm run build`
- `npm test`

Final test result:

- `tests 97`
- `pass 97`
- `fail 0`
- `cancelled 0`
- `skipped 0`

The new tests increased coverage from the previous test count and specifically verify the confidence validation behavior introduced by this PR.

---

## Review Notes

The key design decision in this PR is that confidence validation happens before writing to `pythCache`.

This is intentional because the cache is part of the trusted Pyth price path. Rejecting unsafe Pyth prices before caching prevents low-confidence oracle data from being reused later as if it were a reliable Pyth source.

The helper module is intentionally pure and small so reviewers can inspect the validation logic independently from the keeper runtime.

This PR is focused only on fixing issue #67 and avoids unrelated refactors.

## Summary by CodeRabbit

* **New Features**
  * Added a new configuration option for tuning Pyth price confidence checks.
  * Pyth price updates now use stricter confidence-based validation before being accepted.
* **Bug Fixes**
  * Improved rejection of invalid, missing, or overly uncertain Pyth prices.
  * Added safeguards for timestamp and confidence edge cases.
* **Tests**
  * Added coverage for confidence threshold handling and invalid price data scenarios.

### CODE REVIEW

File: src/pyth-confidence.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+  // Pyth uses the same exponent for price and confidence, so rawConf/rawPrice
+  // gives the same percentage ratio without precision loss from scaling.
+  const confidencePct = Math.abs(rawConf / rawPrice) * 100;
+
+  if (!Number.isFinite(price) || !Number.isFinite(conf) || !Number.isFinite(confidencePct)) {
+    return null;
+  }
+
+  return { price, conf, confidencePct };
```
[the remark this answers is not in the document]
Comment (Bayyan16) [reply src/pyth-confidence.ts:49]: I added an explicit scale underflow guard before returning parsed Pyth prices, and added regression coverage for exponent underflow plus non-finite confidence input.

**Comment by dcccrypto** [comment]
Thank you for this. Closing, with the reasoning below.

**The change itself is sound** — fail-closed on a wide or malformed Pyth confidence interval is
the right default, and `src/pyth-confidence.ts` is a clean pure helper.

**But the live keeper never reads Pyth at all.** It prices from mainnet DEX pools;
`src/cross-cluster/price-reader.ts:591` says so in its own comment — *"Same DEX-pool source as
readPoolPriceE6 (**no Pyth**)"*. The `fetchPythPrices` path this guards exists only in
`src/index.ts`.

**The wiring has no coverage.** The 7 new tests exercise `parsePythPriceQuality` directly and do
pass on their own merits — but I reverted `src/index.ts` to `origin/main` (removing the entire
call site, restoring the exact bug) and **all 33 tests still passed**. Nothing asserts that the
helper is called, or that it is called *before* `pythCache.set`, which is the load-bearing
property. `isPythConfidenceAcceptable` is exported but used only by tests.

---

### Why "dead path" is the deciding fact

The running keeper is **`src/cross-cluster.ts`** — launchd (`com.percolator.oracle-keeper`, live
now) → `start-keeper.sh` → `exec npm run cross-cluster:live` → `tsx src/cross-cluster.ts`,
confirmed against `ps -o command=` on the running process. A full transitive import closure from
that entry reaches **13 modules, and `src/index.ts` is not one of them**. Nothing in `src/`
imports `./index`.

`Dockerfile:40` *is* `CMD ["npx","tsx","src/index.ts"]`, so a container would run it — but
`percolator-oracle-keeper-production`, `oracle-keeper-production` and
`percolator-keeper-production` on Railway all return `404 Application not found`. There is no
deployment where this file executes.

### This is not a blanket rejection

Two of your PRs were **adopted today**, with your commits cherry-picked unmodified and your
authorship intact:

- **#94 → #96** — the mark-smoother gap guard. Your tests there were real; I ran the negative
  control and they fail against unpatched source.
- **#89 → #97** — `rpc-url.ts`, rewired onto the live entry point where it now actually protects
  the Helius API key in `MAINNET_RPC_URL`/`DEVNET_RPC_URL`.

If you want to re-file any of this against `src/cross-cluster/`, it would be very welcome — that
is where changes reach production. Worth pairing each with a negative control (revert the fix,
watch the new test fail); several PRs in this batch passed with the fix fully reverted.
